lib/compat/wordpress-6.0/class-wp-theme-json-gutenberg.php

```diff
@@ -119,6 +166,81 @@ protected function get_block_classes( $style_nodes ) {
 		return $block_rules;
 	}
 
+	/**
+	 * Given a styles array, it extracts the style properties
+	 * and adds them to the $declarations array following the format:
+	 *
+	 * ```php
+	 * array(
+	 *   'name'  => 'property_name',
+	 *   'value' => 'property_value,
+	 * )
+	 * ```
+	 *
+	 * @param array $styles Styles to process.
+	 * @param array $settings Theme settings.
+	 * @param array $properties Properties metadata.
+	 * @param string $selector Selector for styles.
+	 * @return array Returns the modified $declarations.
+	 */
+	protected static function compute_style_properties( $styles, $settings = array(), $properties = null, $selector = null ) {
+		if ( null === $properties ) {
+			$properties = static::PROPERTIES_METADATA;
+		}
+
+		$declarations = array();
+		$root_variable_duplicates = array();
+
+		if ( empty( $styles ) ) {
+			return $declarations;
+		}
+
+		foreach ( $properties as $css_property => $value_path ) {
+			$value = static::get_property_value( $styles, $value_path );
+
+			if ( strpos( $css_property, '--wp--style--root--') === 0 && static::ROOT_BLOCK_SELECTOR !== $selector ) {			
+				continue;
+			}
+
+			if ( strpos( $css_property, '--wp--style--root--') === 0 ) {
+				$root_variable_duplicates[] = substr( $css_property, strlen('--wp--style--root--') );
+			}
+
+			// Look up protected properties, keyed by value path.
+			// Skip protected properties that are explicitly set to `null`.
+			if ( is_array( $value_path ) ) {
+				$path_string = implode( '.', $value_path );
+				if (
+					array_key_exists( $path_string, static::PROTECTED_PROPERTIES ) &&
+					_wp_array_get( $settings, static::PROTECTED_PROPERTIES[ $path_string ], null ) === null
+				) {
+					continue;
+				}
+			}
+
+			// Skip if empty and not "0" or value represents array of longhand values.
+			$has_missing_value = empty( $value ) && ! is_numeric( $value );
+			if ( $has_missing_value || is_array( $value ) ) {
+				continue;
+			}
+
+			$declarations[] = array(
+				'name'  => $css_property,
+				'value' => $value,
+			);
+		}
+
+		// If a variable value is added to the root, the corresponding property should be removed.
+		foreach ( $root_variable_duplicates as $duplicate ) {
```

Comment:
I've been playing with an alternative for [block gap](https://github.com/WordPress/gutenberg/pull/39870), in which I've created a second array `const ROOT_PROPERTIES_METADATA` that contains all the possible root properties.

In this case it would look like this:

```php
const ROOT_PROPERTIES_METADATA = array(
                // all the rest here...
                // except "padding", "padding-top" etc, which are replaced by the following CSS vars
		'--wp--style--root--padding'              => array( 'spacing', 'padding' ),
		'--wp--style--root--padding-top'       => array( 'spacing', 'padding', 'top' ),
		'--wp--style--root--padding-right'     => array( 'spacing', 'padding', 'right' ),
		'--wp--style--root--padding-bottom' => array( 'spacing', 'padding', 'bottom' ),
		'--wp--style--root--padding-left'        => array( 'spacing', 'padding', 'left' ),
	);
```

Then the only change to `compute_style_properties()` method would be:

```php
		if ( null === $properties ) {
			$properties =  static::ROOT_BLOCK_SELECTOR === $selector ? static::ROOT_PROPERTIES_METADATA : static::PROPERTIES_METADATA;
		}
```

Just noting for reference! It might be completely bonkers.

Seems to work for padding at least 😆 but haven't battle tested it yet. 🤔

Reply:
Ooooh that's a great idea! Much neater logic.